lib/rule-tester/rule-tester.js

```diff
@@ -304,6 +305,40 @@ function throwForbiddenMethodError(methodName, prototype) {
     };
 }
 
+/**
+ * Extracts names of {{ placeholders }} from the reported message.
+ * @param   {string} message Reported message
+ * @returns {string[]} Array of placeholder names
+ */
+function getMessagePlaceholders(message) {
+    const matcher = getPlaceholderMatcher();
+
+    return Array.from(message.matchAll(matcher), ([, name]) => name.trim());
+}
+
+/**
+ * Returns the placeholders in the reported messages but
+ * only includes the placeholders available in the raw message if data is not specified or
+ * excludes introduced placeholders specified in the data values.
+ * @param {string} message The reported message
+ * @param {string} raw The raw message specified in the rule meta.messages
+ * @param {undefined|Record<unknown, unknown>} data The passed
+ * @returns {string[]} Missing placeholder names
+ */
+function getMissingMessagePlaceholders(message, raw, data) {
+    const existing = getMessagePlaceholders(message);
+
+    if (data === void 0) {
+        const available = getMessagePlaceholders(raw);
+
+        return existing.filter(name => available.includes(name));
+    }
+
+    const introduced = new Set(Object.values(data).flatMap(value => (typeof value === "string" ? getMessagePlaceholders(value) : [])));
+
+    return existing.filter(name => !introduced.has(name));
```

Comment:
```suggestion
    let existing = getMessagePlaceholders(message);

    //  no potential placeholders were found in the reported message, so there's nothing further to check
    if (existing.length === 0) {
        return existing;
    }

    const original = getMessagePlaceholders(raw);

    // filter out those that don't appear in the original template message as they're certainly not placeholders
    existing = existing.filter(name => original.includes(name));

    if (data) {
        const provided = Object.keys(data);

        // since the hydrated message texts are compared, we can ignore those for which the data is provided in the test
        existing = existing.filter(name => !provided.includes(name));
    }

    return existing;
```

I think this check would be more robust if we always exclude those that don't appear in the meta message and those for which the data is provided in the test.

Reply:
I refactored this method in another way and also kept the case for adding placeholders via data properties (see `no-restricted-syntax` case).